File: evaluation/S1/cpp/Hard/2181-count-the-number-of-ideal-arrays.cpp
// Source: https://leetcode.com/problems/count-the-number-of-ideal-arrays/   |   Difficulty: Hard
//
// Problem Description:
// You are given two integers n and maxValue, which are used to describe an ideal array.
//
// A 0-indexed integer array arr of length n is considered ideal if the following conditions hold:
//
//
// 	Every arr[i] is a value from 1 to maxValue, for 0 <= i < n.
// 	Every arr[i] is divisible by arr[i - 1], for 0 < i < n.
//
//
// Return the number of distinct ideal arrays of length n. Since the answer may be very large, return it modulo 109 + 7.
//
// Example:
// Input: n = 2, maxValue = 5
// Output: 10
// Explanation: The following are the possible ideal arrays:
// - Arrays starting with the value 1 (5 arrays): [1,1], [1,2], [1,3], [1,4], [1,5]
// - Arrays starting with the value 2 (2 arrays): [2,2], [2,4]
// - Arrays starting with the value 3 (1 array): [3,3]
// - Arrays starting with the value 4 (1 array): [4,4]
// - Arrays starting with the value 5 (1 array): [5,5]
// There are a total of 5 + 2 + 1 + 1 + 1 = 10 distinct ideal arrays.
//
// Constraints:
// 2 <= n <= 104
// 	1 <= maxValue <= 104
//
// Helpful references (internal KB):
// - Linear Sieve (array, sieve)
//   • When to use: Use when you need to find the smallest prime factor or full prime factorization for all numbers up to a given limit N in linear time. It's particularly useful for tasks requiring precomputed factorizations.
//   • Idea: The Linear Sieve computes the smallest prime factor for every number up to N, ensuring each composite number is marked exactly once by its smallest prime factor. It achieves this in O(N) time complexity.
//   • Invariants: Every number `i` from 2 to N has its `lp[i]` correctly set to its smallest prime factor.; The `primes` array contains all prime numbers less than or equal to the current iteration `i`.
//   • Tips: Stores the smallest prime factor (lp) for each number.; Efficiently generates all prime numbers up to N.
//   • Pitfalls: Higher memory usage compared to the classic Sieve of Eratosthenes due to storing lp array.; Can be slower in practice than optimized Sieve of Eratosthenes for merely generating primes.
// - Modular Multiplicative Inverse (array, number, prefix-sum, mod-exp)
//   • When to use: Use when needing to compute modular inverses for multiple numbers in an array or a range, especially when efficiency is critical, by leveraging prefix and suffix products to reduce inverse computations.
//   • Idea: This technique efficiently computes modular inverses for multiple numbers by transforming individual inverse calculations into a product of prefix, suffix, and a single overall inverse. It achieves O(N + log M) complexity for an array or O(M) for a range.
//   • Invariants: For 0 <= i < N, prefix_products[i] holds (arr[0] * ... * arr[i-1]) % m.; For 0 <= i < N, suffix_products[i] holds (arr[i+1] * ... * arr[N-1]) % m.
//   • Tips: Precompute prefix products and suffix products of the array elements modulo m.; Calculate the modular inverse of the total product of all elements once using modular exponentiation.
//   • Pitfalls: Ensure all numbers for which an inverse is sought are coprime to the modulus m.; Watch for off-by-one errors when calculating prefix and suffix product indices.
// - Non-negative Integers without Consecutive Ones (number, enumeration, simulation)
//   • When to use: Use when counting numbers up to a given limit that satisfy a specific bit pattern constraint, particularly when the limit is small enough for direct iteration and checking.
//   • Idea: This approach iterates through every non-negative integer from 0 up to the given limit num. For each integer, it simulates checking its binary representation to determine if it contains any two consecutive '1's.
//   • Invariants: count stores the number of integers k (where 0 <= k < i) that do not have consecutive ones.; All integers j such that 0 <= j < i have been processed.
//   • Tips: Use bitwise operations like (i & (i >> 1)) to efficiently detect consecutive ones.; Remember to include 0 in the count as it satisfies the condition (no consecutive ones).
//   • Pitfalls: Performance degradation for large num due to linear scan and per-number bit check.; Incorrectly handling the upper bound of the iteration (e.g., i < num instead of i <= num).
//
// Notes:
// - Keep the given function signature unchanged.
// - Use standard C++ only; do NOT rely on non-standard extensions or platform-specific APIs.
// - Do not change input/output formats or return value semantics.
// - Conformance: Prefer modern standard C++ (C++11/14/17) features that are portable; avoid compiler-specific extensions (e.g., __int128 unless necessary and supported).
// - If you introduce any helper functions, include their full implementations in this file; no hidden helpers.

class Solution {
public:
    int idealArrays(int n, int maxValue) {
        const int MOD = 1000000007;

        // Precompute factorials and inverse factorials for combinations
        std::vector<long long> fact(n + 20, 1), invFact(n + 20, 1);
        for (int i = 2; i < fact.size(); ++i) {
            fact[i] = fact[i - 1] * i % MOD;
        }
        invFact.back() = modInverse(fact.back(), MOD);
        for (int i = invFact.size() - 2; i >= 0; --i) {
            invFact[i] = invFact[i + 1] * (i + 1) % MOD;
        }

        auto comb = [&](int a, int b) {
            if (b > a || b < 0) return 0LL;
            return fact[a] * invFact[b] % MOD * invFact[a - b] % MOD;
        };

        // Linear Sieve to find smallest prime factors
        std::vector<int> lp(maxValue + 1, 0);
        std::vector<int> primes;
        for (int i = 2; i <= maxValue; ++i) {
            if (lp[i] == 0) {
                lp[i] = i;
                primes.push_back(i);
            }
            for (int p : primes) {
                if (p > lp[i] || p * i > maxValue) break;
                lp[p * i] = p;
            }
        }

        long long result = 0;
        for (int val = 1; val <= maxValue; ++val) {
            std::unordered_map<int, int> primeCount;
            int x = val;
            while (x > 1) {
                primeCount[lp[x]]++;
                x /= lp[x];
            }

            long long ways = 1;
            for (const auto& [prime, count] : primeCount) {
                ways = ways * comb(n - 1 + count, count) % MOD;
            }
            result = (result + ways) % MOD;
        }

        return static_cast<int>(result);            
    }
};

File: evaluation/S0/python/Hard/2181-count-the-number-of-ideal-arrays.py
# Source: https://leetcode.com/problems/count-the-number-of-ideal-arrays/   |   Difficulty: Hard
#
# Problem Description:
# You are given two integers n and maxValue, which are used to describe an ideal array.
#
# A 0-indexed integer array arr of length n is considered ideal if the following conditions hold:
#
#
# 	Every arr[i] is a value from 1 to maxValue, for 0 <= i < n.
# 	Every arr[i] is divisible by arr[i - 1], for 0 < i < n.
#
#
# Return the number of distinct ideal arrays of length n. Since the answer may be very large, return it modulo 109 + 7.
#
# Example:
# Input: n = 2, maxValue = 5
# Output: 10
# Explanation: The following are the possible ideal arrays:
# - Arrays starting with the value 1 (5 arrays): [1,1], [1,2], [1,3], [1,4], [1,5]
# - Arrays starting with the value 2 (2 arrays): [2,2], [2,4]
# - Arrays starting with the value 3 (1 array): [3,3]
# - Arrays starting with the value 4 (1 array): [4,4]
# - Arrays starting with the value 5 (1 array): [5,5]
# There are a total of 5 + 2 + 1 + 1 + 1 = 10 distinct ideal arrays.
#
# Constraints:
# 2 <= n <= 104
# 	1 <= maxValue <= 104
#

class Solution(object):
    def idealArrays(self, n, maxValue):
        """
        :type n: int
        :type maxValue: int
        :rtype: int
        """
        MOD = 10**9 + 7

        # Precompute factorials and inverse factorials for combinations
        max_comb = n + 14  # since maxValue <= 10^4, the max number of prime factors is small
        fact = [1] * (max_comb + 1)
        inv_fact = [1] * (max_comb + 1)

        for i in range(2, max_comb + 1):
            fact[i] = fact[i - 1] * i % MOD

        inv_fact[max_comb] = pow(fact[max_comb], MOD - 2, MOD)
        for i in range(max_comb - 1, 0, -1):
            inv_fact[i] = inv_fact[i + 1] * (i + 1) % MOD

        def comb(a, b):
            if a < b or b < 0:
                return 0
            return fact[a] * inv_fact[b] % MOD * inv_fact[a - b] % MOD

        # Sieve to find prime factors
        spf = list(range(maxValue + 1))  # smallest prime factor
        for i in range(2, int(maxValue**0.5) + 1):
            if spf[i] == i:
                for j in range(i * i, maxValue + 1, i):
                    if spf[j] == j:
                        spf[j] = i

        def prime_factors_count(x):
            count = {}
            while x > 1:
                p = spf[x]
                count[p] = count.get(p, 0) + 1
                x //= p
            return count

        result = 0
        for val in range(1, maxValue + 1):
            pf_count = prime_factors_count(val)
            ways = 1
            for exp in pf_count.values():
                ways = ways * comb(n - 1 + exp, exp) % MOD
            result = (result + ways) % MOD

        return result